File: ranreal_sparse.py
# -*- coding: utf-8 -*-

from monta_grafo import Grafo

class RanRealSparse:
    def __init__(self, caminho):
        self._caminho = caminho
        self.grafo_instancia = Grafo()
        self.num_arestas = 0
        #faz a leitura do arquivo da instancia
        with open(caminho) as instancia:
            i=0
            for line in instancia.readlines():
                ls = line.rstrip()
                # print(ls)
                if i==0:
                    # Lendo a primeira linha
                    linha1 = ls.split(' ')
                    # print(linha1)
                    self.numero_elementos = linha1[0]
                    self.numero_clusters = linha1[1]
                    # Obtem limites inferior e superior da soma dos pesos dos nós
                    self.L = linha1[3]
                    self.U = linha1[4]
                    # Obtem os pesos dos nós
                    lista_pesos_nos = linha1[linha1.index('W')+1:]
                    # Cria os nós no grafo com os pesos
                    for i in range(len(lista_pesos_nos)):
                        self.grafo_instancia.add_no(i, lista_pesos_nos[i])
                else:
                    # Cria aresta com peso
                    no = ls.split(' ')[0]
                    no_vizinho = ls.split(' ')[1]
                    peso = ls.split(' ')[2]
                    self.grafo_instancia.add_aresta(no, no_vizinho, peso)
                    self.num_arestas += 1
                i+=1

    def getGrafo(self):
        return self.grafo_instancia

    def get_L(self):
        return self.L

    def get_U(self):
        return self.U

    def __str__(self) -> str:
        return '--Instância--\nnumero_elementos: {}\nnumero_clusters: {}\nnum_arestas: {}\nL: {}\nU: {}'.format(
            self.numero_elementos, self.numero_clusters, self.num_arestas, self.L, self.U
        )


# Teste, comentar quando terminar de desenvolver
# inst = RanRealSparse('instancias/Sparse82/Sparse82_01.txt')
# print(inst)

File: ranreal_sparse.js
const Grafo = require('./grafo')
const fs = require('fs')

/**
 * Classe onde é realizada a leitura das instâncias RanReal e Sparse
 */
class RanRealSparse {

    numArestas = 0
    numeroElementos = 0
    numeroClusters = 0

    /**
     * Construtor da classe, onde é fornecido o caminho da instância
     * @param {string} caminho 
     */
    constructor(caminho) {

        this.grafoInstancia = new Grafo()

        //faz a leitura do arquivo da instancia
        let dataBuffer = fs.readFileSync(caminho);
        const instancia = dataBuffer.toString().split('\n');

        let i = 0;
        for (const linha of instancia) {
            if (i === 0) {
                // Lendo a primeira linha
                const linha1 = linha.trim().split(' ')

                // Obtendo dados do problema
                this.numeroElementos = parseInt(linha1[0])
                this.numeroClusters = parseInt(linha1[1])

                // Obtem limites inferior e superior da soma dos pesos dos nós
                this.L = parseInt(linha1[3])
                this.U = parseInt(linha1[4])

                // Obtem os pesos dos nós
                const lista_pesos_nos = linha1.slice(linha1.indexOf('W') + 1)

                // Cria os nós no grafo com os pesos
                for (let j = 0; j < lista_pesos_nos.length; j++) {
                    this.grafoInstancia.addNo(j, parseFloat(lista_pesos_nos[j]))
                }
            } else {
                // Cria aresta com peso
                const no = parseInt(linha.split(' ')[0])
                const no_vizinho = parseInt(linha.split(' ')[1])
                const peso = parseFloat(linha.split(' ')[2])

                // Se a aresta tem peso 0, ela não vai influenciar no resultado final
                // então, ela não é incluida no grafo final da instância
                if (peso !== 0 && !isNaN(no) && !isNaN(no_vizinho)) {
                    this.grafoInstancia.addAresta(no, no_vizinho, peso)
                    this.numArestas += 1
                }
            }
            i++;
        }
    }

    getNumClusters() {
        return this.numeroClusters
    }

    getNumElementos() {
        return this.numeroElementos
    }

    getGrafo() {
        return this.grafoInstancia
    }

    get_L() {
        return this.L
    }

    get_U() {
        return this.U
    }

}

module.exports = RanRealSparse; 